### PR TITLE
StringArraySelector Feature and Persistence Improvements(Speed, ThreadSafe)

### PR DESCRIPTION
**Pull request by StanleyFung**
ShiftStringArraySelector allows you to specify an array of values and then pick one to be the selected value by providing an index. 

Noticed that persistence was getting very slow with many ShiftValues and realized there is no need to open and close the database for every transaction after some research. Found an updated example from creator of SnappyDB here: https://github.com/theappbusiness/kc-android/blob/master/app/src/main/java/com/nabilhachicha/kc/data/DatabaseImpl.java,
and refactored our persistence. It is now thread safe opening ShiftMenu is much faster. 

You will notice now that we do not call close() on the db at any point(like in the example). I did not notice any irregular behaviour, exceptions, or errors, but am asking if it is necessary here: https://github.com/nhachicha/SnappyDB/issues/64.

### CODE REVIEW

File: shift/src/main/java/org/coursera/android/shift/ShiftPref.java [hunk 1]
```diff
@@ -13,14 +13,38 @@
...
-    void setValue(T value);
+    public ShiftPref(ShiftPersistenceManager persistenceManager,
+                     String key,
+                     T defaultValue,
+                     Class<T> myClass) {
+        PERSISTENCE = persistenceManager;
+        KEY = key;
+        DEFAULT_VALUE = defaultValue;
+        CLASS = myClass;
+        if (!isValueSet()) {
```
Comment (StanleyFung) [inline shift/src/main/java/org/coursera/android/shift/ShiftPref.java:30]: @BenSchwab I added this because otherwise every time you quit the application and relaunch, it will use the default values instead of the ones you have saved.

File: shift/src/main/java/org/coursera/android/shift/ShiftPref.java [hunk 2]
```diff
@@ -13,14 +13,38 @@
...
-interface ShiftPref<T> {
-    T getValue();
+abstract class ShiftPref <T> {
+    protected final ShiftPersistenceManager PERSISTENCE;
+    protected final String KEY;
+    protected final T DEFAULT_VALUE;
+    protected final Class<T> CLASS;
 
-    void setValue(T value);
+    public ShiftPref(ShiftPersistenceManager persistenceManager,
```
Comment (StanleyFung) [inline shift/src/main/java/org/coursera/android/shift/ShiftPref.java:22]: Turned ShiftPref into an abstract class so I could reuse code for isValueSet, delete, and setValueToDefault

File: shift/src/main/java/org/coursera/android/shift/StringListSelector.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+limitations under the License.
+ */
+
+package org.coursera.android.shift;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+class StringListSelector {
```
Comment (StanleyFung) [inline shift/src/main/java/org/coursera/android/shift/StringListSelector.java:20]: At first I tried having extending a generic ListSelector class but discovered that SnappyDB's kryo serialization can not handle the generic type and although it can insert it into the database, it can not properly retrieve it. To make other types of ListSelectors (int, float, etc.) we will have to keep this format.

File: README.md [hunk 4]
```diff
@@ -23,7 +23,7 @@ Shift continues to be immensely helpful for building the Coursera Android applic
 
 ####Gradle
 ``` java
-compile 'org.coursera.android:shift:1.0.1'
+compile 'org.coursera.android:shift:1.0.2'
```
Comment (BenSchwab) [inline README.md:26]: Should this be 1.0.3?

**Comment by StanleyFung** [comment]
StringArraySelector and ArraySelector are actually not used, I will delete them.

File: README.md (hunk 4 above)
Comment (StanleyFung) [inline README.md:26]: Was gonna make a separate commit for changing the version once this is merged and I reupload to jcenter but yea it should be 1.0.3.

**Comment by BenSchwab** [comment]
LGTM after merge conflicts are resolved
